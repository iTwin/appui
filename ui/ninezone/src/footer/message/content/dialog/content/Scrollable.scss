/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
@import "@bentley/ui-core/lib/colors";
@import "../../../../../_scrollbar";
@import "../_variables";

.nz-footer-message-content-dialog-content-scrollable {
  height: 100%;
  padding-bottom: $nz-dialog-padding;
  box-sizing: border-box;

  > .nz-content {
    height: 100%;
    padding-left: $nz-dialog-padding;
    padding-top: $nz-dialog-padding;
    padding-right: $nz-dialog-padding;
    box-sizing: border-box;
    overflow: auto;

    @include nz-scrollbar($color: $background-4);
  }
}
