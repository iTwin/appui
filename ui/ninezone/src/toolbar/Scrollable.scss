/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

@import "_variables";

@mixin scrollTransition {
  transition-property: margin;
  transition-duration: 0.3s;
  transition-timing-function: ease-in-out;
}

.nz-toolbar-scrollable {
  $desktop-space-between-toolbar-items: 1px;

  position: relative;

  > .nz-items-viewport {
    height: 100%;
    width: 100%;
    position: relative;
    overflow: hidden;

    > .nz-items {
      @include scrollTransition;
    }
  }

  > .nz-history {
    > .nz-viewport {
      overflow: hidden;
      position: relative;

      > .nz-container {
        position: relative;

        @include scrollTransition;

        > .nz-item {
          position: relative;

          > div > * {
            pointer-events: auto;
          }
        }
      }
    }
  }

  > .nz-panels {
    @include scrollTransition;
  }

  &.nz-vertical {
    > .nz-history {
      > .nz-viewport {
        > .nz-container {
          > .nz-item {
            height: $nz-desktop-item-height;
            margin-top: $nz-desktop-space-between-toolbar-items;
          }
        }
      }
    }
  }

  &.nz-horizontal {
    > .nz-history {
      > .nz-viewport {
        > .nz-container {
          display: grid;
          grid-auto-flow: column;
          left: 1px;

          > .nz-item {
            width: $nz-desktop-item-width;
            margin-left: $nz-desktop-space-between-toolbar-items;
          }
        }
      }
    }
  }

  > .nz-items-viewport {
    > .nz-left, > .nz-right {
      position: absolute;

      &:not(.nz-is-visible) {
        display: none;
      }
    }
  }

  &.nz-vertical {
    width: $nz-desktop-item-height;

    &.nz-scroll-indicator-left-only {
      > .nz-items-viewport {
        > .nz-left {
          height: $nz-desktop-item-width - $nz-items-border-width - $desktop-space-between-toolbar-items;
        }
      }
    }

    &.nz-scroll-indicator-right-only {
      > .nz-right {
        height: $nz-desktop-item-width - $nz-items-border-width - $desktop-space-between-toolbar-items;
      }
    }

    > .nz-items-viewport {
      > .nz-left, > .nz-right {
        width: 100%;
        height: $nz-desktop-item-height;
      }

      > .nz-left {
        top: 0;
      }

      > .nz-right {
        bottom: 0;
      }
    }
  }

  &.nz-horizontal {
    height: $nz-desktop-item-height;

    &.nz-scroll-indicator-left-only {
      > .nz-items-viewport {
        > .nz-left {
          width: $nz-desktop-item-width - $nz-items-border-width - $desktop-space-between-toolbar-items;
        }
      }
    }

    &.nz-scroll-indicator-right-only {
      > .nz-right {
        width: $nz-desktop-item-width - $nz-items-border-width - $desktop-space-between-toolbar-items;
      }
    }

    > .nz-items-viewport {
      > .nz-left, > .nz-right {
        top: 0;
        height: 100%;
        width: $nz-desktop-item-width;
      }

      > .nz-left {
        left: 0;
      }

      > .nz-right {
        right: 0;
      }

      > .nz-items {
        display: inline-flex;
      }
    }
  }
}
