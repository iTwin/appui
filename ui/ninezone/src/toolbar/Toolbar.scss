/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

@import "../_z-index";
@import "_variables";

.nz-toolbar-toolbar {
  display: inline-block;
  position: relative;

  > .nz-panels {
    pointer-events: none;

    @include nz-z-index(toolbarPanel);

    > .nz-item {
      > * {
        pointer-events: auto;
      }
    }
  }

  > .nz-history {
    pointer-events: none;

    @include nz-z-index(toolbarHistory);

    > .nz-item > div {
      > * {
        pointer-events: auto;
      }
    }
  }

  > .nz-expanded {
    position: absolute;
  }

  &.nz-vertical {
    > .nz-expanded {
      > .nz-item {
        display: grid;
        height: $nz-desktop-item-height;
        margin-top: $nz-desktop-space-between-toolbar-items;
      }
    }

    &.nz-panel-alignment-start {
      > .nz-panels {
        > .nz-item {
          align-content: start;
        }
      }
    }

    &.nz-panel-alignment-end {
      > .nz-panels {
        > .nz-item {
          align-content: end;
        }
      }
    }
  }

  &.nz-horizontal {
    > .nz-expanded {
      display: flex;

      > .nz-item {
        width: $nz-desktop-item-width;
        margin-left: $nz-desktop-space-between-toolbar-items;
        display: grid;
        justify-content: center;
      }
    }

    > .nz-panels {
      > .nz-item {
        > * {
          float: left;
        }
      }
    }

    &.nz-panel-alignment-start {
      > .nz-panels {
        > .nz-item {
          justify-content: start;
        }
      }
    }

    &.nz-panel-alignment-end {
      > .nz-panels {
        > .nz-item {
          justify-content: end;
        }
      }
    }
  }

  &.nz-direction-left {
    > .nz-expanded {
      left: $nz-items-border-width;
      transform: translateX(-100%);
    }
  }

  &.nz-direction-right {
    > .nz-expanded {
      left: 100%;
      transform: translateX(-$nz-items-border-width);
    }
  }

  &.nz-direction-bottom {
    > .nz-expanded {
      top: 100%;
      transform: translateY(-$nz-items-border-width);
    }
  }
}
