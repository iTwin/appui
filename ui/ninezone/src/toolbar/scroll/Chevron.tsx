/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Toolbar */

import * as classnames from "classnames";
import * as React from "react";
import CommonProps, { NoChildrenProps } from "../../utilities/Props";
import Direction, { DirectionHelpers } from "../../utilities/Direction";
import "./Chevron.scss";

/** Properties of [[Chevron]] component. */
export interface ChevronProps extends CommonProps, NoChildrenProps {
  /** Direction of chevron. */
  direction: Direction;
  /** Function called when chevron is clicked. */
  onClick?: () => void;
}

/** Chevron component used in [[Indicator]]. */
export default class Chevron extends React.Component<ChevronProps> {
  public render() {
    const className = classnames(
      "nz-toolbar-scroll-chevron",
      DirectionHelpers.getCssClassName(this.props.direction),
      this.props.className);

    return (
      <div
        className={className}
        style={this.props.style}
        onClick={this.props.onClick}
      />
    );
  }
}
