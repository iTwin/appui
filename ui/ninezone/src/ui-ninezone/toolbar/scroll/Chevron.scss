/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

@import "../../_geometry";
@import "_variables";

.nz-toolbar-scroll-chevron {
  $color: currentColor;

  cursor: pointer;

  &.nz-direction-left {
    @include triangle-left($scroll-indicator-height, $scroll-indicator-width, $color);
  }

  &.nz-direction-right {
    @include triangle-right($scroll-indicator-height, $scroll-indicator-width, $color);
  }

  &.nz-direction-top {
    @include triangle-top($scroll-indicator-width, $scroll-indicator-height, $color);
  }

  &.nz-direction-bottom {
    @include triangle-bottom($scroll-indicator-width, $scroll-indicator-height, $color);
  }
}
