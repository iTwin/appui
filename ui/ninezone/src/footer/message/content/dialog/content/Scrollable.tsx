/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Message */

import * as classnames from "classnames";
import * as React from "react";
import { CommonProps, NoChildrenProps } from "../../../../../utilities/Props";
import "./Scrollable.scss";

/** Properties of [[ScrollableContent]] component. */
export interface ScrollableContentProps extends CommonProps, NoChildrenProps {
  /** Actual content. */
  content?: React.ReactNode;
}

/** Scrollable content of [[Dialog]] component. */
export class ScrollableContent extends React.PureComponent<ScrollableContentProps> {
  public render() {
    const className = classnames(
      "nz-footer-message-content-dialog-content-scrollable",
      this.props.className);

    return (
      <div
        className={className}
        style={this.props.style}
      >
        <div className="nz-content">
          {this.props.content}
        </div>
      </div>
    );
  }
}
