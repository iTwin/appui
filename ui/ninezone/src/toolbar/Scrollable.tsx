/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
/** @module Toolbar */

import * as classnames from "classnames";
import * as React from "react";
import Direction, { OrthogonalDirection } from "../utilities/Direction";
import Indicator from "./scroll/Indicator";
import Toolbar, { ToolbarProps } from "./Toolbar";
import Items from "./Items";
import "./Scrollable.scss";

/** Properties of [[Scrollable]] component. See [[ScrollableDefaultProps]] */
export interface ScrollableProps extends ToolbarProps {
  /** Describes number of visible elements. Should not be lower than 3. */
  visibleItemThreshold?: number;
  /** Function called when component scrolls. */
  onScroll?: () => void;
}

/** Default properties of [[ScrollableProps]] used in [[Scrollable]] component. */
export interface ScrollableDefaultProps extends ScrollableProps {
  /** Defaults to 5. */
  visibleItemThreshold: number;
}

/** State of [[Scrollable]] component. */
export interface ScrollableState {
  /** Describes component scroll offset. */
  scrollOffset: number;
}

/** A [[Toolbar]] with scroll overflow strategy. */
export default class Scrollable extends React.Component<ScrollableProps, ScrollableState> {
  private static readonly _DESKTOP_ITEM_WIDTH = 40;
  private static readonly _DESKTOP_ITEM_HEIGHT = Scrollable._DESKTOP_ITEM_WIDTH;
  private static readonly _BORDER_WIDTH = 1;

  public static readonly defaultProps: Partial<ScrollableDefaultProps> = {
    visibleItemThreshold: 5,
  };

  public readonly state: Readonly<ScrollableState> = {
    scrollOffset: 0,
  };

  /** @returns True if props is [[ScrollableDefaultProps]] */
  public static isWithDefaultProps(props: ScrollableProps): props is ScrollableDefaultProps {
    if (props.visibleItemThreshold === undefined)
      return false;
    return true;
  }

  private getVisibleItemCount() {
    if (!Scrollable.isWithDefaultProps(this.props))
      throw new TypeError();
    const threshold = this.props.visibleItemThreshold;
    const itemCount = Toolbar.getItemCount(this.props);
    return Math.min(threshold, itemCount);
  }

  private getViewportLength(itemLength: number) {
    let length = this.getVisibleItemCount() * itemLength;
    if (this.isLeftMostScrolled())
      length += Scrollable._BORDER_WIDTH;
    if (this.isRightMostScrolled())
      length += Scrollable._BORDER_WIDTH;
    return length;
  }

  private getViewportStyle(direction: OrthogonalDirection) {
    switch (direction) {
      case OrthogonalDirection.Vertical: {
        const verticalStyle: React.CSSProperties = {
          height: this.getViewportLength(Scrollable._DESKTOP_ITEM_HEIGHT),
        };
        return verticalStyle;
      }
      case OrthogonalDirection.Horizontal: {
        const horizontalStyle: React.CSSProperties = {
          width: this.getViewportLength(Scrollable._DESKTOP_ITEM_WIDTH),
        };
        return horizontalStyle;
      }
    }
  }

  private getHistoryViewportLength(itemLength: number) {
    let length = this.getVisibleItemCount() * itemLength;
    if (this.isLeftScrollIndicatorVisible()) {
      length -= itemLength;
    }
    if (this.isRightScrollIndicatorVisible())
      length -= itemLength;
    return length;
  }

  private getHistoryViewportOffset(itemLength: number) {
    let top = 0;
    if (this.isLeftScrollIndicatorVisible()) {
      top += itemLength - Scrollable._BORDER_WIDTH;
    }
    if (!this.isLeftMostScrolled() && this.isRightMostScrolled()) {
      top -= Scrollable._BORDER_WIDTH;
    }
    return top;
  }

  private getHistoryViewportStyle(direction: OrthogonalDirection): React.CSSProperties {
    switch (direction) {
      case OrthogonalDirection.Vertical: {
        const height = this.getHistoryViewportLength(Scrollable._DESKTOP_ITEM_HEIGHT);
        const top = this.getHistoryViewportOffset(Scrollable._DESKTOP_ITEM_HEIGHT);
        return {
          height,
          top,
        };
      }
      case OrthogonalDirection.Horizontal: {
        const width = this.getHistoryViewportLength(Scrollable._DESKTOP_ITEM_WIDTH);
        const left = this.getHistoryViewportOffset(Scrollable._DESKTOP_ITEM_WIDTH);
        return {
          width,
          left,
        };
      }
    }
  }

  private getHistoryScrolledStyle(direction: OrthogonalDirection): React.CSSProperties {
    const isLeftScrollIndicatorVisible = this.isLeftScrollIndicatorVisible();
    switch (direction) {
      case OrthogonalDirection.Vertical: {
        const marginTop = this.state.scrollOffset * -Scrollable._DESKTOP_ITEM_HEIGHT;
        let top = 0;
        if (isLeftScrollIndicatorVisible) {
          top -= Scrollable._DESKTOP_ITEM_HEIGHT;
        }
        return {
          marginTop,
          top,
        };
      }
      case OrthogonalDirection.Horizontal: {
        const marginLeft = this.state.scrollOffset * -Scrollable._DESKTOP_ITEM_WIDTH;
        let left = 0;
        if (isLeftScrollIndicatorVisible) {
          left -= Scrollable._DESKTOP_ITEM_WIDTH;
        }
        return {
          marginLeft,
          left,
        };
      }
    }
  }

  private getScrollOffset(itemLength: number) {
    let offset = this.state.scrollOffset * itemLength;
    if (this.isLeftScrollIndicatorVisible())
      offset += Scrollable._BORDER_WIDTH;
    if (!this.isLeftMostScrolled() && this.isRightMostScrolled())
      offset += Scrollable._BORDER_WIDTH;
    return -offset;
  }

  private getScrolledStyle(direction: OrthogonalDirection) {
    switch (direction) {
      case OrthogonalDirection.Vertical: {
        const marginTop = this.getScrollOffset(Scrollable._DESKTOP_ITEM_HEIGHT);
        const verticalStyle: React.CSSProperties = { marginTop };
        return verticalStyle;
      }
      case OrthogonalDirection.Horizontal: {
        const marginLeft = this.getScrollOffset(Scrollable._DESKTOP_ITEM_WIDTH);
        const horizontalStyle: React.CSSProperties = { marginLeft };
        return horizontalStyle;
      }
    }
  }

  private _handleLeftScroll = () => {
    this.setState(
      (prevState) => {
        let scrollOffset = prevState.scrollOffset - 1;
        scrollOffset = Math.max(0, scrollOffset);
        return {
          ...prevState,
          scrollOffset,
        };
      },
      () => this.props.onScroll && this.props.onScroll(),
    );
  }

  private _handleRightScroll = () => {
    this.setState(
      (prevState, props) => {
        const itemCnt = Toolbar.getItemCount(props);
        let scrollOffset = prevState.scrollOffset + 1;
        scrollOffset = Math.min(itemCnt - this.getVisibleItemCount(), scrollOffset);
        return {
          ...prevState,
          scrollOffset,
        };
      },
      () => this.props.onScroll && this.props.onScroll(),
    );
  }

  private isLeftScrollIndicatorVisible() {
    return this.state.scrollOffset > 0;
  }

  private isRightScrollIndicatorVisible() {
    const itemCnt = Toolbar.getItemCount(this.props);
    return itemCnt - this.getVisibleItemCount() - this.state.scrollOffset > 0;
  }

  private isLeftMostScrolled() {
    return this.state.scrollOffset === 0;
  }

  private isRightMostScrolled() {
    const itemCnt = Toolbar.getItemCount(this.props);
    return itemCnt - this.getVisibleItemCount() - this.state.scrollOffset === 0;
  }

  public render() {
    const isLeftScrollIndicatorVisible = this.isLeftScrollIndicatorVisible();
    const isRightScrollIndicatorVisible = this.isRightScrollIndicatorVisible();
    const direction = Toolbar.getToolbarDirection(this.props);

    const { className, ...props } = this.props;
    const scrollableClassName = classnames(
      "nz-toolbar-scrollable",
      isLeftScrollIndicatorVisible && !isRightScrollIndicatorVisible && "nz-scroll-indicator-left-only",
      !isLeftScrollIndicatorVisible && isRightScrollIndicatorVisible && "nz-scroll-indicator-right-only",
      className);

    const leftScrollIndicatorClassName = classnames(
      "nz-left",
      isLeftScrollIndicatorVisible && "nz-is-visible",
    );
    const rightScrollIndicatorClassName = classnames(
      "nz-right",
      isRightScrollIndicatorVisible && "nz-is-visible",
    );

    const viewportStyle = this.getViewportStyle(direction);
    const scrolledStyle = this.getScrolledStyle(direction);
    return (
      <Toolbar
        className={scrollableClassName}
        renderItems={(items) => (
          <div
            className="nz-items-viewport"
            style={viewportStyle}
          >
            <Items
              className="nz-items"
              direction={direction}
              style={scrolledStyle}
            >
              {items}
            </Items>
            <Indicator
              className={leftScrollIndicatorClassName}
              direction={direction === OrthogonalDirection.Vertical ? Direction.Top : Direction.Left}
              onClick={this._handleLeftScroll}
            />
            <Indicator
              className={rightScrollIndicatorClassName}
              direction={direction === OrthogonalDirection.Vertical ? Direction.Bottom : Direction.Right}
              onClick={this._handleRightScroll}
            />
          </div>
        )}
        renderHistoryItems={(historyItems) => (
          <div
            className="nz-expanded nz-history"
          >
            <div
              className="nz-viewport"
              style={this.getHistoryViewportStyle(direction)}
            >
              <div
                className="nz-container"
                style={this.getHistoryScrolledStyle(direction)}
              >
                {historyItems}
              </div>
            </div>
          </div>
        )}
        renderPanelItems={(panelItems) => (
          <div
            className="nz-expanded nz-panels"
            style={scrolledStyle}
          >
            {panelItems}
          </div>
        )}
        {...props}
      />
    );
  }
}
