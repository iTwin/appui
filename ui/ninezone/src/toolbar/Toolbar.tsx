/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Toolbar */

import * as classnames from "classnames";
import * as React from "react";

import Direction, { DirectionHelpers, OrthogonalDirection, OrthogonalDirectionHelpers } from "../utilities/Direction";
import CommonProps, { NoChildrenProps, FlattenChildren } from "../utilities/Props";
import { ExpandableItemProps } from "./item/expandable/Expandable";
import Items from "./Items";
import "./Toolbar.scss";

interface ToolbarItem {
  item: React.ReactNode;
  panel: React.ReactNode;
  history: React.ReactNode;
}

/** Available alignment modes of [[Toolbar]] panels. */
export enum ToolbarPanelAlignment {
  Start,
  End,
}

export class ToolbarPanelAlignmentHelpers {
  /** Class name of [[ToolbarPanelAlignment.Start]] */
  public static readonly START_CLASS_NAME = "nz-panel-alignment-start";
  /** Class name of [[ToolbarPanelAlignment.End]] */
  public static readonly END_CLASS_NAME = "nz-panel-alignment-end";

  /** @returns Class name of specified [[ToolbarPanelAlignment]] */
  public static getCssClassName(panelAlignment: ToolbarPanelAlignment): string {
    switch (panelAlignment) {
      case ToolbarPanelAlignment.Start:
        return ToolbarPanelAlignmentHelpers.START_CLASS_NAME;
      case ToolbarPanelAlignment.End:
        return ToolbarPanelAlignmentHelpers.END_CLASS_NAME;
    }
  }
}

/** Properties of [[Toolbar]] component. */
export interface ToolbarProps extends CommonProps, NoChildrenProps {
  /** Describes to which direction the history/panel items are expanded. Defaults to: [[Direction.Bottom]] */
  expandsTo?: Direction;
  /** Items of the toolbar. I.e. [[ExpandableItem]], [[Icon]], [[Item]], [[Overflow]] */
  items?: React.ReactNode;
  /** Describes how expanded panels are aligned. Defaults to: [[ToolbarPanelAlignment.Start]] */
  panelAlignment?: ToolbarPanelAlignment;
  /** Function called to render history items. */
  renderHistoryItems?: (historyItems: React.ReactNode) => React.ReactNode;
  /** Function called to render items. */
  renderItems?: (items: React.ReactNode) => React.ReactNode;
  /** Function called to render panel items. */
  renderPanelItems?: (panelItems: React.ReactNode) => React.ReactNode;
}

/**
 * A toolbar that may contain items.
 * @note See [[Scrollable]] for toolbar with scroll overflow strategy.
 */
export default class Toolbar extends React.Component<ToolbarProps> {
  /** @returns Toolbar direction based on [[ToolbarProps.expandsTo]] */
  public static getToolbarDirection(props: ToolbarProps): OrthogonalDirection {
    const expandsTo = Toolbar.getExpandsTo(props);
    const orthogonalDirection = DirectionHelpers.getOrthogonalDirection(expandsTo);
    return OrthogonalDirectionHelpers.inverse(orthogonalDirection);
  }

  /** @returns Count of toolbar items. */
  public static getItemCount(props: ToolbarProps) {
    const items = FlattenChildren(props.items);
    return React.Children.count(items);
  }

  /** @returns True if item is [[ExpandableItem]] */
  private static isExpandableItem(item: React.ReactChild): item is React.ReactElement<ExpandableItemProps> {
    if (React.isValidElement<ExpandableItemProps>(item))
      return true;
    return false;
  }

  private static getExpandsTo(props: ToolbarProps) {
    return props.expandsTo === undefined ? Direction.Bottom : props.expandsTo;
  }

  private static getPanelAlignment(props: ToolbarProps) {
    return props.panelAlignment === undefined ? ToolbarPanelAlignment.Start : props.panelAlignment;
  }

  private getToolbarItems(): ToolbarItem[] {
    const items = FlattenChildren(this.props.items);

    const toolbarItems = React.Children.map<ToolbarItem>(items, (child, index) => {
      const key = React.isValidElement(child) && child.key ? child.key : index;

      const panelRef = React.createRef<HTMLDivElement>();
      const panel = (
        <div
          key={key}
          className="nz-item"
          ref={panelRef}
        >
        </div>
      );

      const historyRef = React.createRef<HTMLDivElement>();
      const history = (
        <div
          key={key}
          className="nz-item"
          ref={historyRef}
        >
        </div>
      );

      const item = !Toolbar.isExpandableItem(child) ? child :
        React.cloneElement<ExpandableItemProps>(child, {
          key,
          renderHistoryTo: () => {
            if (!historyRef.current)
              throw new ReferenceError();
            return historyRef.current;
          },
          renderPanelTo: () => {
            if (!panelRef.current)
              throw new ReferenceError();
            return panelRef.current;
          },
        });

      return {
        item,
        panel,
        history,
      };
    });

    if (!toolbarItems)
      return [];
    return toolbarItems;
  }

  public render() {
    const orthogonalDirection = Toolbar.getToolbarDirection(this.props);
    const expandsTo = Toolbar.getExpandsTo(this.props);
    const panelAlignment = Toolbar.getPanelAlignment(this.props);
    const className = classnames(
      "nz-toolbar-toolbar",
      DirectionHelpers.getCssClassName(expandsTo),
      OrthogonalDirectionHelpers.getCssClassName(orthogonalDirection),
      ToolbarPanelAlignmentHelpers.getCssClassName(panelAlignment),
      this.props.className);

    const toolbarItems = this.getToolbarItems();
    const items = toolbarItems.map((toolbarItem) => toolbarItem.item);
    const historyItems = toolbarItems.map((toolbarItem) => toolbarItem.history);
    const panelItems = toolbarItems.map((toolbarItem) => toolbarItem.panel);
    return (
      <div
        className={className}
        style={this.props.style}
      >
        {this.props.renderHistoryItems ? this.props.renderHistoryItems(historyItems) :
          <div
            className="nz-expanded nz-history"
          >
            {historyItems}
          </div>
        }
        {this.props.renderPanelItems ? this.props.renderPanelItems(panelItems) :
          <div
            className="nz-expanded nz-panels"
          >
            {panelItems}
          </div>
        }
        {this.props.renderItems ? this.props.renderItems(items) :
          <Items
            className="nz-items"
            direction={orthogonalDirection}
          >
            {items}
          </Items>
        }
      </div >
    );
  }
}
