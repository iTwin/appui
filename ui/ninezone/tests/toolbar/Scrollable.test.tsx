/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import { mount, shallow } from "enzyme";
import * as React from "react";
import Scrollable, { ScrollableState } from "../../src/toolbar/Scrollable";
import { Direction } from "../../src/utilities/Direction";
import Chevron from "../../src/toolbar/scroll/Chevron";
import { ToolbarPanelAlignment } from "../../src/toolbar/Toolbar";

// tslint:disable-next-line:variable-name
const Item: React.StatelessComponent = () => {
  return <div />;
};

describe("<Scrollable />", () => {
  it("should render", () => {
    mount(<Scrollable />);
  });

  it("renders correctly", () => {
    shallow(<Scrollable />).should.matchSnapshot();
  });

  it("renders with visible item threshold", () => {
    shallow(<Scrollable visibleItemThreshold={2} />).should.matchSnapshot();
  });

  it("renders with expandsTo", () => {
    shallow(<Scrollable expandsTo={Direction.Left} />).should.matchSnapshot();
  });

  it("renders with panelAlignment", () => {
    shallow(<Scrollable panelAlignment={ToolbarPanelAlignment.End} />).should.matchSnapshot();
  });

  it("renders vertical with overflow scrolled left correctly", () => {
    const sut = shallow(
      <Scrollable
        visibleItemThreshold={3}
        items={
          <>
            <Item />
            <Item />
            <Item />
            <Item />
          </>
        }
      />,
    );
    sut.setState({
      scrollOffset: 1,
    });
    sut.should.matchSnapshot();
  });

  it("renders vertical with overflow scrolled right most correctly", () => {
    const sut = shallow(
      <Scrollable
        visibleItemThreshold={2}
        items={
          <>
            <Item />
            <Item />
            <Item />
            <Item />
          </>
        }
      />,
    );
    sut.setState({
      scrollOffset: 2,
    });
    sut.should.matchSnapshot();
  });

  it("initial scroll offset should be 0", () => {
    const sut = shallow<Scrollable, ScrollableState>(<Scrollable />);
    sut.state().scrollOffset.should.eq(0);
  });

  it("should handle left scroll indicator click events", () => {
    const sut = mount<Scrollable, ScrollableState>(<Scrollable visibleItemThreshold={3} />);
    sut.setState({
      scrollOffset: 2,
    });
    const scroll = sut.findWhere((node) => node.name() === "div" && node.hasClass("nz-left"));
    scroll.exists().should.true;

    const indicator = scroll.findWhere((node) => node.type() === Chevron);
    indicator.exists().should.true;

    indicator.simulate("click");
    sut.state().scrollOffset.should.eq(1);
  });

  it("should handle right scroll indicator click events", () => {
    const sut = mount<Scrollable, ScrollableState>(
      <Scrollable
        visibleItemThreshold={2}
        items={
          <>
            <Item />
            <Item />
            <Item />
            <Item />
            <Item />
          </>
        }
      />,
    );
    sut.setState({
      scrollOffset: 1,
    });
    const scroll = sut.findWhere((node) => node.name() === "div" && node.hasClass("nz-right"));
    scroll.exists().should.true;

    const indicator = scroll.findWhere((node) => node.type() === Chevron);
    indicator.exists().should.true;

    indicator.simulate("click");
    sut.state().scrollOffset.should.eq(2);
  });
});
