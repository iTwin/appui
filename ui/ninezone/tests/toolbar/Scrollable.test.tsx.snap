// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Scrollable /> renders correctly 1`] = `
<Toolbar
  className="nz-toolbar-scrollable"
  expandsTo={0}
  renderHistoryItems={[Function]}
  renderItems={[Function]}
  renderPanelItems={[Function]}
  visibleItemThreshold={5}
/>
`;

exports[`<Scrollable /> renders vertical with overflow scrolled left correctly 1`] = `
<Toolbar
  className="nz-toolbar-scrollable nz-scroll-indicator-left-only"
  expandsTo={0}
  items={
    <UNDEFINED>
      <div />
      <div />
      <div />
      <div />
    </UNDEFINED>
  }
  renderHistoryItems={[Function]}
  renderItems={[Function]}
  renderPanelItems={[Function]}
  visibleItemThreshold={3}
/>
`;

exports[`<Scrollable /> renders vertical with overflow scrolled right most correctly 1`] = `
<Toolbar
  className="nz-toolbar-scrollable nz-scroll-indicator-left-only"
  expandsTo={0}
  items={
    <UNDEFINED>
      <div />
      <div />
      <div />
      <div />
    </UNDEFINED>
  }
  renderHistoryItems={[Function]}
  renderItems={[Function]}
  renderPanelItems={[Function]}
  visibleItemThreshold={2}
/>
`;

exports[`<Scrollable /> renders with expandsTo 1`] = `
<Toolbar
  className="nz-toolbar-scrollable"
  expandsTo={2}
  renderHistoryItems={[Function]}
  renderItems={[Function]}
  renderPanelItems={[Function]}
  visibleItemThreshold={5}
/>
`;

exports[`<Scrollable /> renders with visible item threshold 1`] = `
<Toolbar
  className="nz-toolbar-scrollable"
  expandsTo={0}
  renderHistoryItems={[Function]}
  renderItems={[Function]}
  renderPanelItems={[Function]}
  visibleItemThreshold={2}
/>
`;
